File: cs50/pset1/credit/credit.c
//A program to check if a credit card number is valid.

#include <stdio.h>
#include <cs50.h>
#include <math.h>

int luhn(long long n);

int main(void)
{
    long long cardNum = get_long_long("Enter card number: ");

    int digits = log10(cardNum) + 1;
    int firstDigit = cardNum / pow(10, (digits - 1));
    int secondDigit = cardNum / pow(10, (digits - 2));
    secondDigit %= 10;

    //Amex cards begin with 34 or 37 and have 15 digits
    if (firstDigit == 3 && (secondDigit == 4 || secondDigit == 7) && digits == 15)
    {
        if (luhn(cardNum) == 0)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    //Mastercards begin with 51 through 55 and have 16 digits
    else if (firstDigit == 5 && secondDigit > 0 && secondDigit <= 5 && digits == 16)
    {
        if (luhn(cardNum) == 0)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    //Visa cards begin with 4 and have 13 or 16 digits
    else if (firstDigit == 4 && (digits == 13 || digits == 16))
    {
        if (luhn(cardNum) == 0)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    else
    {
        printf("INVALID\n");
    }
}

//Function uses Luhn's algorithm to check number validity.
//If returns zero, card is valid. Non-zero, invalid.
int luhn(long long n)
{
    int arr[16] = {0, 0, 0, 0, 0};
    int luhnTotal = 0;

    //Create array of digits from long long "n"
    for (int i = 0; i < 16; i++)
    {
        arr[15 - i] = n % 10;
        n /= 10;
    }

    //Start at end. Add last digit, multiply second to last by two then add, repeat.
    for (int k = 0; k < 16; k++)
    {
        if (k % 2 == 0)
        {
            luhnTotal += arr[15 - k];
        }
        else
        {
            if ((arr[15 - k] * 2) > 9)
            {
                luhnTotal += (arr[15 - k] * 2) - 9; //If product is > 9, Luhn's Algorithm says to subtract 9
            }
            else
            {
                luhnTotal += (arr[15 - k] * 2);
            }
        }
    }
    return (luhnTotal % 10);
}

File: cs50/pset6/credit/credit.py
# A program to check if a credit card number is valid.

from cs50 import get_int
from math import log10


def main():
    """
    Get a credit card number from the user.
    Check the # of digits and first two digits of card to
    pre-check validity and card type.
    If good, call function to run Luhn's algorithm on number
    If Luhn = True, print card type.
    Else, print INVALID
    """

    card_num = get_int("Enter card number: ")
    if card_num < 10**12 or card_num > 10**16:
        print("INVALID")

    digits = int(log10(card_num)) + 1
    first_digit = card_num // pow(10, (digits - 1))
    second_digit = (card_num // pow(10, (digits - 2))) % 10

    # Amex cards begin with 34 or 37 and have 15 digits
    if (digits == 15
            and first_digit == 3
            and (second_digit == 4
                 or second_digit == 7)):
        if (luhn(card_num, digits)):
            print("AMEX")

    # Mastercards begin with 51 through 55 and have 16 digits
    elif (digits == 16
            and first_digit == 5
            and second_digit > 0
            and second_digit <= 5):
        if (luhn(card_num, digits)):
            print("MASTERCARD")

    # Visa cards begin with 4 and have 13 or 16 digits
    elif ((digits == 13
            or digits == 16)
            and first_digit == 4):
        if (luhn(card_num, digits)):
            print("VISA")

    else:
        print("INVALID")


def luhn(card_num, digits):
    """
    Runs Luhn's algorithm on a pre-processed card number.
    """
    arr = []

    # Starting from right-most digit, build array
    for i in range(digits):
        # d is value of current digit
        d = card_num % 10

        # if place marker is even (from right), insert d into array
        if i % 2 == 0:
            arr.insert(0, d)

        # if place marker is odd (from right), insert d*2 into array
        else:
            # if d*2 > 9, Luhn says subtract 9
            if 2 * d > 9:
                arr.insert(0, (2 * d - 9))
            else:
                arr.insert(0, (2 * d))
        card_num //= 10

    total = sum(arr)

    # Luhn >> sum of array % 10 should equal zero
    if sum(arr) % 10 == 0:
        return True
    else:
        return False


if __name__ == '__main__':
    main()